docstring/csc108_docstring.py

```diff
@@ -1,288 +1,94 @@
 
-    
-def trim_docstring_down_to_type_contract(full_docstring):
-    """Extracts the line we want (the type contract) from the entire docstring.
-    This assumes there always will be a docstring, whereby it will read the top
-    line and continue reading until it hits an empty line, of which it will then
-    assume the docstring header to be done. This allows for a docstring to be
-    multiple lines, like:
-        (int, set of list of tuple of(int, str), obj)
-            ->
-                tuple of (int, int)
-    This is needed since PEP-8 being enforced in CSC108 means the possibility of
-    the docstring being greater than 80 characters, and it must be supported if
-    its too long.
-    This function will properly extract the string and compress it into a single
-    string with all newlines/improper spacing removed.
-    Note: This does not check for correctness of the docstring header contract,
-    its job is only to extract the docstring, and let the parser error out if
-    the type contract is malformed.
-    """
-    lines = full_docstring.split('\n')  # TODO - Support \r\n too.
-    merged_docstring = ''
-    for line in lines:
-        stripped_line = line.strip()
-        if len(stripped_line) <= 0:
-            break
-        # Lack of whitespace between lines could cause problems, so we will
-        # assume there is supposed to be a space to replace the new line.
-        # Since the parser will skip any extra space, this is safe to do.
-        merged_docstring += stripped_line + ' '
-    return merged_docstring[:-1]  # Trim last space (TODO - do this better...)
+element.define(set_parser | list_parser | dict_parser | tuple_parser | any_class)
+elements.define(maybe(many(element + skip_str(','))) + element)
 
+# TODO - Need to strip a [] leftover if there's one element at the beginning.
+docstring_parser = skip_str('(') + elements + skip_str(')') + skip_str('->') + element
 
-def parse_csc108_docstring(full_docstring):
-    """Takes the header line of a docstring which is supposed to be the
-    docstring, and parses it. If parsing fails, a NotA108DocstringException
-    is raised.
-    Returns two nodes as a tuple: (Args tree DocNode, result tree DocNode).
-    """
-    line = trim_docstring_down_to_type_contract(full_docstring)
-    parser = DocstringParser(line)
-    return parser.args_node, parser.return_node
+# -----------------------------------------------------------------------------
+
+# non_right_arrow = some(lambda x: x != '>')
+# single_arrow = a_str('>') + non_right_arrow
+# double_arrow = a_str('>>') + non_right_arrow
+# docstring_description = oneplus(non_right_arrow | single_arrow | double_arrow) >> to_str
+
+# docstring_doctest = a_str('>>>') + (oneplus(some(is_valid_char)) >> to_str)
+
+# -----------------------------------------------------------------------------
+
+docstring_raw = '(dict of {SomeFakeClass, tuple of (bool, int, str, NoneType)}, int) -> list of list of SomeClass'
```

Comment:
Please remove this code (and the commented-out block above it). If you want, you can move the string into the test file.